### PR TITLE
Provider class endpoints

### PR DESCRIPTION
**Pull request by bartekryba**
PR for #33

ContractClassGatewaySerializer will probably be changed, after compressing contracts by cptartur is merged.

commented out DeployTransaction is related to #43

### CODE REVIEW

File: lib/src/main/kotlin/starknet/data/types/transactions.kt [hunk 1]
```diff
@@ -103,18 +103,18 @@ data class DeclareTransaction(
...
+//    val contractAddressSalt: Felt,
+//    val constructorCalldata: Calldata,
+//    val nonce: Felt?
+// ) : Transaction() {
+//    override val type = TransactionType.DEPLOY
+//
+//    override fun getHash(): Felt {
+//        TODO("Not yet implemented")
+//    }
+// }
```
Comment (Solpatium) [inline lib/src/main/kotlin/starknet/data/types/transactions.kt:117]: 👁️

File: lib/src/main/kotlin/starknet/provider/gateway/GatewayProvider.kt [hunk 2]
```diff
@@ -120,6 +120,58 @@ class GatewayProvider(
...
+        TODO("Not available from gateway")
+    }
+
+    override fun getClassAt(blockNumber: Int, contractAddress: Felt): Request<ContractClass> {
+        TODO("Not available from gateway")
+    }
+
+    override fun getClassAt(blockTag: BlockTag, contractAddress: Felt): Request<ContractClass> {
+        TODO("Not available from gateway")
+    }
```
Comment (Solpatium) [inline lib/src/main/kotlin/starknet/provider/gateway/GatewayProvider.kt:144]: Only common parts should be a part of Provider interface.

File: lib/src/main/kotlin/starknet/data/types/contracts.kt [hunk 3]
```diff
@@ -38,3 +50,69 @@ data class CompiledContract(
...
+
+    override fun deserialize(decoder: Decoder): ContractClass {
+        val response = ContractClassGateway.serializer().deserialize(decoder)
+        val bos = ByteArrayOutputStream()
+
+        val programString = Json.encodeToString(response.program)
+
+        GZIPOutputStream(bos).bufferedWriter(UTF_8).use { writer ->
+            writer.write(programString)
+        }
```
Comment (Solpatium) [inline lib/src/main/kotlin/starknet/data/types/contracts.kt:91]:
```suggestion
        GZIPOutputStream(bos).bufferedWriter(UTF_8).apply { write(programString) }
```

**Review by Solpatium — CHANGES_REQUESTED** [review]

File: lib/src/main/kotlin/starknet/data/types/transactions.kt (hunk 1 above)
Comment (bartekryba) [reply lib/src/main/kotlin/starknet/data/types/transactions.kt:117]: Well, I didn't want to remove it here, as there is an issue for getting rid of data class duplication (#43) and I think it's better to get rid of these classes there.

Comment (Solpatium) [reply lib/src/main/kotlin/starknet/data/types/transactions.kt:117]: But you commented it out here anyway, so there is a change. Why do we need a commented version of it here?

**Review by Solpatium — APPROVED** [review]

File: lib/src/test/kotlin/starknet/provider/ProviderTest.kt [hunk 4]
```diff
@@ -133,6 +134,145 @@ class ProviderTest {
...
+        if (provider is JsonRpcProvider) {
+            return
+        }
+
+        val request = provider.getClassHashAt(BlockTag.LATEST, contractAddress)
+        val response = request.send()
+
+        assertNotNull(response)
+
+        print(response)
```
Comment (cptartur) [inline lib/src/test/kotlin/starknet/provider/ProviderTest.kt:243]: I think this is not needed

File: lib/src/main/kotlin/starknet/data/hex.kt [hunk 5]
```diff
@@ -12,3 +12,4 @@ fun parseHex(value: String): BigInteger {
 }
 
 fun toHex(value: BigInteger): String = "0x" + value.toString(16)
+fun toHex(value: Int): String = "0x" + value.toString(16)
```
Comment (cptartur) [inline lib/src/main/kotlin/starknet/data/hex.kt:15]: These could probably work as extensions methods, but that may be an issue for possible future refactors.

File: lib/src/main/kotlin/starknet/data/types/contracts.kt [hunk 6]
```diff
@@ -38,3 +50,69 @@ data class CompiledContract(
...
+}
+
+@Serializable
+data class ContractClass(
+    val program: String,
+    @SerialName("entry_points_by_type") val entryPointsByType: EntryPointsByType
+) {
+    @Serializable
+    data class EntryPointsByType(
+        @SerialName("CONSTRUCTOR") val constructor: List<ContractEntryPoint>,
```
[the remark this answers is not in the document]
Comment (bartekryba) [reply lib/src/main/kotlin/starknet/data/types/contracts.kt:70]: Yeah, I agree that it looks better. I wonder if we could add a custom ktlint rule for that.

File: lib/src/main/kotlin/starknet/data/types/contracts.kt [hunk 7]
```diff
@@ -38,3 +50,69 @@ data class CompiledContract(
     // TODO: Add entrypoints, once they are defined
     val program: Program,
 )
+
+@Serializable
+data class ContractEntryPoint(
+    val offset: String,
```
[the remark this answers is not in the document]
Comment (bartekryba) [reply lib/src/main/kotlin/starknet/data/types/contracts.kt:56]: I thought about using some other number type here, as Felt's restraints are unrelated to this value.

[the remark this answers is not in the document]
Comment (cptartur) [reply lib/src/main/kotlin/starknet/data/types/contracts.kt:56]: Maybe we could define some other type then or use BigInteger directly here. Strings aren't very convenient for the end user.

File: lib/src/main/kotlin/starknet/data/hex.kt (hunk 5 above)
Comment (cptartur) [reply lib/src/main/kotlin/starknet/data/hex.kt]: Duplicate

[the remark this answers is not in the document]
Comment (Solpatium) [reply lib/src/main/kotlin/starknet/data/hex.kt]: How are extension methods translated to java? I think `toHex` functions might be used by library users.

[the remark this answers is not in the document]
Comment (Solpatium) [reply lib/src/main/kotlin/starknet/data/hex.kt]: It compiles to 
```
   @NotNull
   public static final String toHex(@NotNull BigInteger $this$toHex) {
      Intrinsics.checkNotNullParameter($this$toHex, "$this$toHex");
      return "0x" + $this$toHex.toString(16);
   }
```
The parameter name is not java friendly.

[the remark this answers is not in the document]
Comment (cptartur) [reply lib/src/main/kotlin/starknet/data/hex.kt]: In #49 I've started adding extensions methods into separate `extensions` module, that way they don't pollute the namespace much. If we keep this as part of the "private api" - don't document them, I think it's fine even if they compile to this in java. If some user decides to use them anyway, they'd be on their own.

File: lib/src/main/kotlin/starknet/data/types/contracts.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (bartekryba) [reply lib/src/main/kotlin/starknet/data/types/contracts.kt]: I decided to use Felt for now

File: lib/src/main/kotlin/starknet/data/types/contracts.kt [hunk 8]
```diff
@@ -71,3 +75,72 @@ data class CompiledContract(
...
+        // FIXME: It doesn't produce the same output as the rpc endpoint
+
+        return ContractClass(program, response.entryPointsByType)
+    }
+
+    override val descriptor: SerialDescriptor
+        get() = ContractClass.serializer().descriptor
+
+    override fun serialize(encoder: Encoder, value: ContractClass) {
+        TODO("Not implemented. It is used only for deserialization.")
```
Comment (cptartur) [inline lib/src/main/kotlin/starknet/data/types/contracts.kt:129]: Perhaps we should throw some exception here?

File: lib/src/main/kotlin/starknet/data/types/contracts.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (cptartur) [reply lib/src/main/kotlin/starknet/data/types/contracts.kt]: Let's convert this into an issue.

**Review by cptartur — APPROVED** [review]
Looks fine to me! There's one unresolved conversation about extensions methods, so I'd resolve it before merging.

File: lib/src/main/kotlin/starknet/data/types/contracts.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (cptartur) [reply lib/src/main/kotlin/starknet/data/types/contracts.kt]: #52

**Pull request merged**